### PR TITLE
Perftest: Split perf test into separeted modules

### PR DESCRIPTION
**Pull request by vasilyMellanox**
## Why
peftest.c file is getting to large and hard to maintain, better split it to several files

### CODE REVIEW

**Comment by vasilyMellanox** [comment]
@yosefe @gleon99

File: src/tools/perf/Makefile.am [hunk 1]
```diff
@@ -11,9 +11,18 @@
...
 
-noinst_HEADERS        = api/libperf.h
+noinst_HEADERS        = \
+	perftest.h \
+	api/libperf.h
+
 bin_PROGRAMS          = ucx_perftest
-ucx_perftest_SOURCES  = perftest.c
+
+ucx_perftest_SOURCES  = \
```
Comment (yosefe) [inline src/tools/perf/Makefile.am:20]: extra space

File: src/tools/perf/lib/libperf_int.h [hunk 2]
```diff
@@ -231,6 +231,47 @@ size_t ucx_perf_get_message_size(const ucx_perf_params_t *params)
     return length;
 }
 
+typedef struct {
+    ucs_status_t (*setup)(ucx_perf_context_t *perf);
+    void         (*cleanup)(ucx_perf_context_t *perf);
+    ucs_status_t (*run)(ucx_perf_context_t *perf);
+    void         (*barrier)(ucx_perf_context_t *perf);
+} ucx_perf_funcs_t;
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_int.h:239]: put structs in beginning of H file

File: src/tools/perf/lib/libperf_int.h [hunk 3]
```diff
@@ -231,6 +231,47 @@ size_t ucx_perf_get_message_size(const ucx_perf_params_t *params)
...
+
+extern ucx_perf_funcs_t ucx_perf_funcs[];
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_int.h:241]: order in H file should be:
1. defines
2. enums
3. structs
4. extern (global vars)
5. func prototypes
6. inline funcs

File: src/tools/perf/lib/libperf_int.h [hunk 4]
```diff
@@ -231,6 +231,47 @@ size_t ucx_perf_get_message_size(const ucx_perf_params_t *params)
...
+
+ucs_status_t ucx_perf_thread_spawn(ucx_perf_context_t *perf,
+                                   ucx_perf_result_t* result);
+
+void ucx_perf_test_prepare_new_run(ucx_perf_context_t *perf,
+                                   const ucx_perf_params_t *params);
+
+void ucx_perf_set_warmup(ucx_perf_context_t* perf,
+                         const ucx_perf_params_t* params);
+
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_int.h:273]: extra space line

File: src/tools/perf/lib/libperf_memory.c [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+#include <tools/perf/lib/libperf_int.h>
+
+#include <string.h>
+#include <unistd.h>
+
+#if _OPENMP
+#   include <omp.h>
+#endif /* _OPENMP */
+
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_memory.c:24]: one more spaceline after

File: src/tools/perf/lib/libperf_thread.c [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        }
+    }
+
+    ucx_perf_thread_report_aggregated_results(perf);
+
+    free(statuses);
+out:
+    return status;
+}
+#else
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_thread.c:144]: space line before and after else

File: src/tools/perf/lib/libperf_thread.c [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+ucs_status_t ucx_perf_thread_spawn(ucx_perf_context_t *perf,
+                                   ucx_perf_result_t* result)
+{
+    ucs_error("Invalid test parameter (thread mode requested without OpenMP capabilities)");
+    return UCS_ERR_INVALID_PARAM;
+}
+#endif /* _OPENMP */
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_thread.c:151]: space line before

File: src/tools/perf/perftest.h [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+#include <getopt.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/poll.h>
+#include <locale.h>
+
+#if defined (HAVE_MPI)
+#  include <mpi.h>
+#elif defined (HAVE_RTE)
+#   include<rte.h>
```
Comment (yosefe) [inline src/tools/perf/perftest.h:31]: space after include

File: src/tools/perf/perftest.h [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    const char                   *desc;
+    const char                   *overhead_lat;
+    unsigned                     window_size;
+} test_type_t;
+
+typedef struct perftest_params {
+    ucx_perf_params_t            super;
+    int                          test_id;
+} perftest_params_t;
+
```
Comment (yosefe) [inline src/tools/perf/perftest.h:68]: extra space

File: src/tools/perf/perftest_params.c [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+    test = &tests[params->test_id];
+
+    if (params->super.max_outstanding == 0) {
+        params->super.max_outstanding = test->window_size;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t read_batch_file(FILE *batch_file, const char *file_name,
```
Comment (yosefe) [inline src/tools/perf/perftest_params.c:294]: should be in perftest_run

File: src/tools/perf/lib/libperf_int.h [hunk 11]
```diff
@@ -231,6 +231,47 @@ size_t ucx_perf_get_message_size(const ucx_perf_params_t *params)
...
+void uct_perf_test_free_mem(ucx_perf_context_t *perf);
+
+ucs_status_t ucx_perf_thread_spawn(ucx_perf_context_t *perf,
+                                   ucx_perf_result_t* result);
+
+void ucx_perf_test_prepare_new_run(ucx_perf_context_t *perf,
+                                   const ucx_perf_params_t *params);
+
+void ucx_perf_set_warmup(ucx_perf_context_t* perf,
+                         const ucx_perf_params_t* params);
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_int.h:272]: are all these functions really need to be in H file?

File: src/tools/perf/perftest.h [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+ucs_status_t adjust_test_params(perftest_params_t *params,
+                                const char *error_prefix);
+
+void print_test_name(struct perftest_context *ctx);
+void print_header(struct perftest_context *ctx);
+
+void print_progress(char **test_names, unsigned num_names,
+                    const ucx_perf_result_t *result, unsigned flags,
+                    int final, int is_server, int is_multi_thread);
```
Comment (yosefe) [inline src/tools/perf/perftest.h:111]: are all these functions really need to be in H file?

File: src/tools/perf/perftest_params.c [hunk 13]
```diff
@@ -0,0 +1,499 @@
+/**
+* Copyright (C) NVIDIA 2021.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "perftest.h"
+
```
Comment (yosefe) [inline src/tools/perf/perftest_params.c:8]: one more space line after

File: src/tools/perf/perftest_print.c [hunk 14]
```diff
@@ -0,0 +1,288 @@
+/**
+* Copyright (C) NVIDIA 2021.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "perftest.h"
+
```
Comment (yosefe) [inline src/tools/perf/perftest_print.c:8]: 1. one more space line after 
2. IMO better unite perftest_print into perftest_run

File: src/tools/perf/perftest_print.c [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+    for (it = UCS_MEMORY_TYPE_HOST; it < UCS_MEMORY_TYPE_LAST; it++) {
+        if (ucx_perf_mem_type_allocators[it] != NULL) {
+            printf("                        %s - %s\n",
+                   ucs_memory_type_names[it],
+                   ucs_memory_type_descs[it]);
+        }
+    }
+}
+
+void usage(const struct perftest_context *ctx, const char *program)
```
Comment (yosefe) [inline src/tools/perf/perftest_print.c:181]: should be in perftest_params

File: src/tools/perf/perftest_run.c [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+    print_header(ctx);
+
+    status = run_test_recurs(ctx, &ctx->params, 0);
+    if (status != UCS_OK) {
+        ucs_error("Failed to run test: %s", ucs_status_string(status));
+    }
+
+    return status;
+}
```
Comment (yosefe) [inline src/tools/perf/perftest_run.c:93]: newline at end of file

**Review by yosefe — COMMENTED** [review]
also pls check CI failures

**Comment by vasilyMellanox** [comment]
@yosefe fixed

File: src/tools/perf/lib/Makefile.am [hunk 17]
```diff
@@ -21,10 +21,11 @@ libucxperf_la_LIBADD   = \
 # C-linkable C++ code - must override any inherited CXXFLAGS
 CXXFLAGS              += -nostdlib $(PERF_LIB_CXXFLAGS) 
 
-noinst_HEADERS = \
-	libperf_int.h
+noinst_HEADERS = libperf_int.h
```
Comment (yosefe) [inline src/tools/perf/lib/Makefile.am:24]: pls revert

File: src/tools/perf/lib/libperf.c [hunk 18]
```diff
@@ -16,12 +16,14 @@
 #include <ucs/arch/bitops.h>
 #include <ucs/sys/module.h>
 #include <ucs/sys/string.h>
-#include <string.h>
+
```
Comment (yosefe) [inline src/tools/perf/lib/libperf.c:19]: remove space line

File: src/tools/perf/lib/libperf.c [hunk 19]
```diff
@@ -1813,21 +1563,13 @@ static void ucp_perf_cleanup(ucx_perf_context_t *perf)
     ucp_cleanup(perf->ucp.context);
 }
 
-static struct {
-    ucs_status_t (*setup)(ucx_perf_context_t *perf);
-    void         (*cleanup)(ucx_perf_context_t *perf);
-    ucs_status_t (*run)(ucx_perf_context_t *perf);
-    void         (*barrier)(ucx_perf_context_t *perf);
-} ucx_perf_funcs[] = {
+ucx_perf_funcs_t ucx_perf_funcs[] = {
```
Comment (yosefe) [inline src/tools/perf/lib/libperf.c:1566]: static?

File: src/tools/perf/perftest.h (hunk 8 above)
Comment (yosefe) [reply src/tools/perf/perftest.h:38]: should be like this
```
#if defined (HAVE_MPI)
# include <mpi.h>
#elif defined (HAVE_RTE)
# include <rte.h>
#endif
```

File: src/tools/perf/perftest.h [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+#include <getopt.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/poll.h>
+#include <locale.h>
+
+#if defined (HAVE_MPI)
+#  include <mpi.h>
+
+#elif defined (HAVE_RTE)
```
Comment (yosefe) [inline src/tools/perf/perftest.h:37]: pls add only the needed includes in this H file (e.g which are needed for stuff defined in H file itself). the rest should go to the relevant C files

File: src/tools/perf/perftest.h (hunk 9 above)
Comment (yosefe) [reply src/tools/perf/perftest.h:76]: pls remove one spaceline

File: src/tools/perf/perftest_params.c (hunk 10 above)
Comment (yosefe) [reply src/tools/perf/perftest_params.c:295]: should be in perftest_run

File: src/tools/perf/perftest_run.c [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+    for (it = UCS_MEMORY_TYPE_HOST; it < UCS_MEMORY_TYPE_LAST; it++) {
+        if (ucx_perf_mem_type_allocators[it] != NULL) {
+            printf("                        %s - %s\n",
+                   ucs_memory_type_names[it],
+                   ucs_memory_type_descs[it]);
+        }
+    }
+}
+
+void usage(const struct perftest_context *ctx, const char *program)
```
Comment (yosefe) [inline src/tools/perf/perftest_run.c:182]: should be in perftest_params

File: src/tools/perf/lib/libperf.c (hunk 19 above)
Comment (vasilyMellanox) [reply src/tools/perf/lib/libperf.c:1566]: no, used in several modules

**Comment by yosefe** [comment]
@vasilyMellanox please avoid force-push and rebase during review because it's not possible to see what has changed since last review.
If there is a conflict please use merge.
Can you please rebase the PR on top of 26d0a6a1c09062f808d6fd81fd613b7b4b55b308 to fix this?

**Comment by vasilyMellanox** [comment]
bot:retest

File: src/tools/perf/perftest.h [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+
+#include "api/libperf.h"
+#include "lib/libperf_int.h"
+
+#if defined (HAVE_MPI)
+#  include <mpi.h>
+
+#elif defined (HAVE_RTE)
+#   include<rte.h>
+
```
Comment (yosefe) [inline src/tools/perf/perftest.h:18]: remove space line

File: src/tools/perf/lib/libperf_int.h [hunk 23]
```diff
@@ -29,13 +29,18 @@ BEGIN_C_DECLS
...
+#define rte_call(_perf, _func, ...) \
+    ((_perf)->params.rte->_func((_perf)->params.rte_group, ## __VA_ARGS__))
+
 
 typedef struct ucx_perf_context        ucx_perf_context_t;
 typedef struct uct_peer                uct_peer_t;
 typedef struct ucp_perf_request        ucp_perf_request_t;
 typedef struct ucx_perf_thread_context ucx_perf_thread_context_t;
 
-
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_int.h:38]: why removed space lines in this file?
we should have 2 space lines in H files?

File: src/tools/perf/lib/libperf_int.h [hunk 24]
```diff
@@ -118,49 +122,75 @@ struct ucx_perf_thread_context {
...
 };
 
+typedef struct {
+    ucs_status_t (*setup)(ucx_perf_context_t *perf);
+    void         (*cleanup)(ucx_perf_context_t *perf);
+    ucs_status_t (*run)(ucx_perf_context_t *perf);
+    void         (*barrier)(ucx_perf_context_t *perf);
+} ucx_perf_funcs_t;
 
-#define UCX_PERF_TEST_FOREACH(perf) \
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_int.h:134]: 👍 for moving to file start

File: src/tools/perf/lib/libperf_int.h [hunk 25]
```diff
@@ -118,49 +122,75 @@ struct ucx_perf_thread_context {
...
+ucs_status_t ucp_perf_test_alloc_mem(ucx_perf_context_t *perf);
+void ucp_perf_test_free_mem(ucx_perf_context_t *perf);
+void ucx_perf_test_memcpy_host(void *dst, ucs_memory_type_t dst_mem_type,
+                               const void *src, ucs_memory_type_t src_mem_type,
+                               size_t count);
+ucs_status_t
+uct_perf_test_alloc_host(const ucx_perf_context_t *perf, size_t length,
+                         unsigned flags, uct_allocated_memory_t *alloc_mem);
+void uct_perf_test_free_host(const ucx_perf_context_t *perf,
+                             uct_allocated_memory_t *alloc_mem);
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_int.h:167]: move ucx_perf_global_init() to libperf_memory.c so these functions could be static and not have to be defined in H file

File: src/tools/perf/lib/libperf_int.h [hunk 26]
```diff
@@ -118,49 +122,75 @@ struct ucx_perf_thread_context {
...
+                                   const ucx_perf_params_t *params);
+void ucx_perf_set_warmup(ucx_perf_context_t* perf,
+                         const ucx_perf_params_t* params);
 
-void uct_perf_barrier(ucx_perf_context_t *perf);
+/**
+ * Get the total length of the message size given by parameters
+ */
+static inline
+size_t ucx_perf_get_message_size(const ucx_perf_params_t *params)
```
Comment (yosefe) [inline src/tools/perf/lib/libperf_int.h:181]: move to libperf_run.c file (as non inline)

File: src/tools/perf/perftest.h [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+#endif
+
+#define MAX_BATCH_FILES         32
```
Comment (yosefe) [inline src/tools/perf/perftest.h:21]: one space line before

File: src/tools/perf/perftest.h [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+    const char                   *overhead_lat;
+    unsigned                     window_size;
+} test_type_t;
+
+typedef struct perftest_params {
+    ucx_perf_params_t            super;
+    int                          test_id;
+} perftest_params_t;
+
+
```
Comment (yosefe) [inline src/tools/perf/perftest.h:56]: remove one space line

File: src/tools/perf/perftest_params.c [hunk 29]
```diff
@@ -0,0 +1,579 @@
...
+    status = init_test_params(&ctx->params);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ctx->server_addr            = NULL;
+    ctx->num_batch_files        = 0;
+    ctx->port                   = 13337;
+    ctx->flags                  = 0;
+    ctx->mpi                    = mpi_initialized;
```
Comment (yosefe) [inline src/tools/perf/perftest_params.c:527]: extra spaces before =

**Review by yosefe — APPROVED** [review]
pls squash (but without rebase)

**Comment by vasilyMellanox** [comment]
bot:retest

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by gleon99** [comment]
@vasilyMellanox , minor: please fix spelling in commit and PR. `separeted` -> `separated`.

**Review by yosefe — APPROVED** [review]

**Pull request merged**